Restructure

flying-sheep opened the pull request:
Big cleanup with the intention to

- [x] Put `core.anndata` on a diet
- [x] Move internal APIs where they belong:

   - [x] centralize `"raw.X"`/`"raw/X"`
   - [x] make `AnnData.__init__` only accept user-facing APIs, …)

- [x] unify names: `isview`→`is_view`, `isbacked`→`filename`
- [x] Do the abstract class thing I always wanted. E.g. there’s only `AxisArrays`, and `AxisArrays(self, 1, **varm)` creates an instance of `AxisArrays.Actual`.

TODO reviews:

- [x] Bring back `AnnData(..., raw=dict(...))` (will be squashed into f430584b967dc7ccc8eadac93fda438527301886)
  - [x] test it
- [x] Rename `*.View`→`*._View`, `*.Actual`→`*._Actual` (will be squashed into 01e96d60ea80e1aee58ebdde28bafe683c46798f)
- [x] introduce more separation of concerns into anndata/sparsedataset/raw

I’m also really hurting to deduplicate `readwrite.{h5ad,zarr}`. What’s the plan @ivirshup?

flying-sheep commented on anndata/core/alignedmapping.py:
```diff
@@ -94,7 +91,7 @@ def isview(self) -> bool:
 
     @property
-    def parent(self) -> "AnnData":
+    def parent(self) -> Union["anndata.AnnData", "raw.Raw"]:
```
I’m unsure what here can have Raw as a parent, and how we should handle this going forward.

Using a AnnData base that supports all the interfaces used by AnnData, Raw, and so on?

ivirshup replied:
Hmm. I can't follow the link from the edit due to the rebase, but I assume this includes `varm`?

flying-sheep replied:
Yes, Raw can have `varm` which is an `AxisArrays`, which means that `parent` can be a `Raw`.

ivirshup replied:
Ah, I think I didn't quite get what you were asking. 

Yeah, I was thinking an `AnnDataBase` or `AbstractAnnData` class would be good. It's a little weird how methods would get split up since raw only has the `var` ones, but it would be good to reuse code.

flying-sheep commented on anndata/core/anndata.py:
```diff
@@ -837,13 +494,15 @@ def _init_as_actual(
-            # is dictionary from reading the file, nothing that is meant for a user
+        elif isinstance(raw, cabc.Mapping):
+            # is dictionary from reading the file,
+            # nothing that is meant for a user
```
Then it shouldn’t be here, but in the function reading from a file. We only want convenience-dict-initialization for things users should be able to supply.

flying-sheep replied:
Addressed in df9fcc6

flying-sheep commented on anndata/core/anndata.py:
```diff
@@ -489,38 +487,10 @@ def _init_as_actual(
-            self._raw = None
-        elif isinstance(raw, Raw):
-            self._raw = raw
-        elif isinstance(raw, cabc.Mapping):
```
Removing the parameter should be fine:

1. `AnnData(raw=None)` is useless for users
2. `AnnData(raw=dict(...))` is an internal API and has to go
3. `AnnData(raw=adata2.raw)` might be OK, we need to evaluate if people use this.

Bonus: Allowing `AnnData(raw=adata2)` would make most sense, as `raw.setter` already only accepts AnnDatas.

But generally, you only want `.raw` if you want to save an AnnData object without losing yout raw data, right? So just having `adata.raw = ...` should be enough.

ivirshup replied:
I think having raw be a dict would be okay as long as it's practically `self.raw = Raw(**raw)`. I think mutable mappings are a pretty useful interface, and I like having this nice recursive definition available to the user.

flying-sheep commented:
<details>
<summary><a href="https://travis-ci.org/theislab/anndata/jobs/617802895">Hmm, this seems like a broken test:</link></summary>

```pytb
______________________ test_write_large_categorical[zarr] ______________________
tmp_path = PosixPath('/tmp/pytest-of-travis/pytest-0/test_write_large_categorical_z0')
diskfmt = 'zarr'
    def test_write_large_categorical(tmp_path, diskfmt):
        M = 30_000
        N = 1000
        ls = np.array(list(ascii_letters))
        cats = np.array(
            sorted(
                "".join(np.random.choice(ls, np.random.choice(range(5, 30))))
                for i in range(int(10_000))
            )
        )
        adata_pth = tmp_path / f"adata.{diskfmt}"
        n_cats = len(np.unique(cats))
        orig = ad.AnnData(
            csr_matrix(([1], ([0], [0])), shape=(M, N)),
            obs=pd.DataFrame(
                dict(
                    cat1=cats[np.random.choice(n_cats, M)],
                    cat2=pd.Categorical.from_codes(
>                       np.random.choice(n_cats, M), cats
                    ),
                )
            ),
        )
anndata/tests/test_readwrite.py:441: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
../../../virtualenv/python3.6.7/lib/python3.6/site-packages/pandas/core/arrays/categorical.py:678: in from_codes
    categories=categories, ordered=ordered, dtype=dtype
../../../virtualenv/python3.6.7/lib/python3.6/site-packages/pandas/core/dtypes/dtypes.py:335: in _from_values_or_dtype
    dtype = CategoricalDtype(categories, ordered)
../../../virtualenv/python3.6.7/lib/python3.6/site-packages/pandas/core/dtypes/dtypes.py:226: in __init__
    self._finalize(categories, ordered, fastpath=False)
../../../virtualenv/python3.6.7/lib/python3.6/site-packages/pandas/core/dtypes/dtypes.py:347: in _finalize
    categories = self.validate_categories(categories, fastpath=fastpath)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
categories = Index(['AABpqcfZHQEVld', 'AAESZRyTyDhuurrUsjriWanAK', 'AAOHJUDaVYKaHOjSkob',
       'AAYEQClnRwDoDZMmXid', 'AApVwUpruN...   'zzpGxxK', 'zzwjQdbdbOGlvAByyptywtofu',
       'zzxmgEbkTWRGvBAHrfGDcBzshkdAB'],
      dtype='object', length=10000)
fastpath = False
    @staticmethod
    def validate_categories(categories, fastpath: bool = False):
        """
        Validates that we have good categories
    
        Parameters
        ----------
        categories : array-like
        fastpath : bool
            Whether to skip nan and uniqueness checks
    
        Returns
        -------
        categories : Index
        """
        from pandas.core.indexes.base import Index
    
        if not fastpath and not is_list_like(categories):
            msg = "Parameter 'categories' must be list-like, was {!r}"
            raise TypeError(msg.format(categories))
        elif not isinstance(categories, ABCIndexClass):
            categories = Index(categories, tupleize_cols=False)
    
        if not fastpath:    
            if categories.hasnans:
                raise ValueError("Categorial categories cannot be null")
    
            if not categories.is_unique:
>               raise ValueError("Categorical categories must be unique")
E               ValueError: Categorical categories must be unique
../../../virtualenv/python3.6.7/lib/python3.6/site-packages/pandas/core/dtypes/dtypes.py:527: ValueError
```

</details>

ivirshup commented:
> I’m also really hurting to deduplicate readwrite.{h5ad,zarr}. What’s the plan @ivirshup?

Pretty sure it works now, and I don't want to mess with it too much before a release. Next version we'll collapse. This may involve upstreaming some stuff to zarr for compatibility, so I think it needs time.

However, I would love to rename `anndata.readwrite` to `anndata.io`

ivirshup commented:
Oh, and on `isbacked`, `filename`, I think these can serve different purposes. `isbacked` makes sense as a way to check whether the object is backed or not, and the other attribute shouldn't be used as a stand-in. 

 At some point wasn't there a "cached" mode where the data was in memory, but `adata.write()` would update the disk version? I think I remember a tutorial having that.

ivirshup commented on anndata/core/aligned_mapping.py:
```diff
@@ -259,11 +278,11 @@ def __init__(
 
-AxisArraysBase._view_class = AxisArraysView
-AxisArraysBase._actual_class = AxisArrays
+AxisArrays.View = _AxisArraysView
```
I think these should be private. It's very possible these classes won't exist in a version or two, so I'd like to expose very little interface.

flying-sheep replied:
Well, everything not documented is private, but sure, let’s make this even clearer with underscores.

ivirshup commented:
Could you ping me when you think this is ready for an in depth look? Or if there are sections you think are done and I should start looking at.

flying-sheep commented:
> Oh, and on `isbacked`, `filename`, I think these can serve different purposes. `isbacked` makes sense as a way to check whether the object is backed or not, and the other attribute shouldn't be used as a stand-in.

Well, its implementation is literally `self.filename is not None`, so either that was wrong or we don’t need `isbacked`.

> Could you ping me when you think this is ready for an in depth look? Or if there are sections you think are done and I should start looking at.

Every commit by itself should address a separate concern, and I consider every commit I didn’t comment on to be ready. Therefore TODO:

- Bring back `AnnData(..., raw=dict(...))` (will be squashed into f430584b967dc7ccc8eadac93fda438527301886)
- Rename `*.View`→`*._View`, `*.Actual`→`*._Actual` (will be squashed into 01e96d60ea80e1aee58ebdde28bafe683c46798f)
- Probably 1 new commit introducing more separation of concerns into anndata/sparsedataset/raw

ivirshup commented on anndata/core/raw.py:
```diff
@@ -16,20 +16,28 @@
-        varm: Union[AxisArrays, AxisArraysView, None] = None,
+        var: Union[
+            AxisArrays, AxisArraysView, Mapping[str, Sequence], None
+        ] = None,
```
`var: Optional[pd.DataFrame] = None`?

flying-sheep replied:
Mapping also works right?

ivirshup replied:
I'd lean no, but do we ever do that? I think `pd.DataFrame` is probably most accurate.

Definitely not `AxisArrays` or `AxisArraysView` though.

flying-sheep replied:
Yeah, it’s tested in `test_base.test_creation`: https://github.com/theislab/anndata/blob/b0e1a004e64d0e3aadbb9265e0a2fd7ead27b416/anndata/tests/test_base.py#L29-L35

ivirshup commented on anndata/core/raw.py:
```diff
@@ -16,20 +16,28 @@
-            self._var = adata.var.copy()
+            self._var = _gen_dataframe(
+                adata.var.copy(), adata.n_vars, ['var_names']
+            )
```
If this is going to come from `adata.var.copy()` (which is a weird thing to happen, and seems wrong), then it would have already been through validation and wouldn't need `_gen_dataframe`,

flying-sheep replied:
Right!

ivirshup replied:
Isn't this the case where it's backed? Why would it copy the parents `.var` when it's backed, but not otherwise?

ivirshup replied:
Haha, I'm probably not the right person to ask about expectations for `.raw`. I do not like that it exists, and think it should go since we can just do `layers` or a separate object.

I think it's behaviour should line up with AnnData as much as possible. Since backed mode didn't really work before (in many cases). I'm hoping we're okay to change this.

I'm wouldn't think that it's possible for `.X` to be specified without `.var` since there should at least be a `var_names`.

flying-sheep commented on anndata/core/sparsedataset.py:
```diff
@@ -375,15 +361,15 @@ def append(self, sparse_matrix):
         indices[orig_data_size:] = sparse_matrix.indices
 
-    def tobacked(self):
+    def to_backed(self) -> BackedSparseMatrix:
```
All private APIs still, so no.

flying-sheep commented on anndata/readwrite/h5ad.py:
```diff
@@ -83,8 +83,8 @@ def write_h5ad(
     filepath = Path(filepath)
-    mode = "a" if adata.isbacked else "w"
-    if adata.isbacked:  # close so that we can reopen below
+    mode = "a" if adata.filename else "w"
```
True, no idea why I put those changes in here. Probably because I was already renaming stuff

flying-sheep commented on anndata/core/file_backing.py:
```diff
@@ -0,0 +1,100 @@
+from os import PathLike
```
Yes:

![grafik](https://user-images.githubusercontent.com/291575/69997157-cdf3d780-1553-11ea-8118-4f056c810a0c.png)

flying-sheep commented on anndata/core/aligned_mapping.py:
```diff
@@ -90,30 +114,30 @@ def axes(self) -> Tuple[int, ...]:
 
-class AlignedViewMixin:
-    parent: "AnnData"
+class AlignedView(AlignedMapping, ABC):
```
They’re ABCs instead of Mixins now. Now PyCharm can e.g. resolve all the variables and autocomplete and so on.

flying-sheep commented on anndata/core/aligned_mapping.py:
```diff
@@ -90,30 +114,30 @@ def axes(self) -> Tuple[int, ...]:
 
-    def copy(self):
-        d = self._actual_class(self.parent, self._axis)
+    def copy(self) -> "_Actual":
```
It is now!

flying-sheep replied:
Ah, sorry, I misunderstood. This is on the base class, therefore this returns an instance of the class variable (which is a subclass of AlignedActual). So this is correct.

ivirshup replied:
Can you point me to the docs for that?

flying-sheep replied:
You’re right, it doesn’t work: python/mypy#8070

ivirshup replied:
I think `-> AlignedView` would be fine for now.

Otherwise, not totally sure off the top of my head. Maybe something like

```python
from typing import Protocol, TypeVar

class View():
    pass

class Actual():
    pass

VT = TypeVar("VT", View)
AT = TypeVar("AT", Actual)

class AlignedMapping(Protocol[AT, VT]):
    _Actual: AT
    _View: VT

    def copy(self) -> AT:
        pass

class Layers("AlignedMapping[_LayersActual, _LayersView]"):
    pass
```

flying-sheep commented on anndata/core/aligned_mapping.py:
```diff
@@ -45,15 +42,42 @@ class AlignedMapping(cabc.MutableMapping, ABC):
+        else:
+            args = self.parent, getattr(self, "_axis", None)
+            kw = self._data
+        return args, kw
```
My switch to `__new__` made pickling fail, `__getnewargs_ex__` fixes that.

But it can indeed be split up!

flying-sheep commented on anndata/core/aligned_mapping.py:
```diff
@@ -273,44 +298,41 @@ class LayersBase(AlignedMapping):
 
 
-class PairwiseArraysBase(AlignedMapping):
+class PairwiseArrays(AlignedMapping, ABC):
```
It’s like `Path`: Calling `PairwiseArrays()` uses `__new__` to instantiate a subclass directly. There will never be a `Path` or `PairwiseArrays` in existence, but it’s the preferred way to get a subclass instance.

ivirshup replied:
`Path` is an abstraction between OSs which gives a consistent interface over different underlying implementations. View and Actual should be functionally different. I think it's a different design pattern for a different use case.

As far as I can tell, none of the AlignedMappings ever use the interface you've described. Currently, I think the `AlignedMapping.__new__` change should be reverted and each class should just use its `__init__`. I don't see a use case, but I do see added indirection.

I think a case which could be useful for this would be to add the `__new__` method to the abstract bases for `Actual` and `View`. Then you could have something like:

```
AlignedMapping(adata, (0,1), vals) -> Layers
AlignedMapping(adata, (0,), vals) -> AxisArrays
AlignedMapping(adata, (0, 0), vals) -> PairwiseArrays
```

But I don't think this needs to happen before a release, and could use a bit more thought.

flying-sheep replied:
The new interface is used. `Layers(...)`, `AxisArrays(...)` or `PairwiseArrays(...)` goes through this single `__new__`:

```py
from .aligned_mapping import AxisArrays, PairwiseArrays, Layers
...
self._layers = Layers(self, vals=(layers or {}))
# it’s now a Layers._Actual instance
```

I did it this way because we could start by just documenting `AxisArrays`, and in case we add new convenience APIs to e.g. `Layers`, we can also document that, without exposing the private APIs that are used to deal with the view/actual distinction.

> I think a case which could be useful for this would be to add the __new__ method to the abstract bases for `Actual` and `View`.

I think `Layers(adata, vals=...)` is already a nice enough API.

ivirshup replied:
> The new interface is used. Layers(...), AxisArrays(...) or PairwiseArrays(...) goes through this single __new__:

But they still all go through their `__init__`? I agree that this code is being used, but it's a change in code with no change to functionality of the API (I don't see changes to where these object are instantiated), which I think should be avoided unless there's and obvious benefit. So far I haven't understood a benefit.

Since this doesn't modify any functionality (at least any functionality I see used), is it something that could happen later?

> I did it this way because we could start by just documenting AxisArrays, and in case we add new convenience APIs to e.g. Layers

How does changing the code for instantiation make this more possible?

flying-sheep replied:
The difference is that one uses `Layers(...)` instead of `LayersActual(...)` or `LayersView(...)` to construct things, i.e. the `Layers._Actual` and `Layers._View` classes are now implementation details. So when documenting the module, we say:

> There’s the ABC `AlignedMapping` with the following properties: …
> 
> There’s the 3 classes AxisArrays, PairwiseArrays, and Layers which are aligned to the axes …, respectively

If we want to expose it, then we can also add:

> Each class can be `.is_view`, with properties … or `not .is_view` with properties …

I think saying there’s 2×3 classes with a common super class is more instructive than saying there’s 6 classes, 3 of which have the superclass `View` and 2 of which have the superclass `LayersBase`…

ivirshup replied:
> The difference is that one uses Layers(...) instead of LayersActual(...) or LayersView(...) to construct things, i.e. the Layers._Actual and Layers._View

Yeah? I don't think that's so different from before, since only `Layers` (which you renamed to `LayersActual`) would ever be constructed outside of the module, and views would be constructed by the `._view` method. I don't any internals here need much more explanation to users than they have now, for layers it's:

> Dictionary-like object with values of the same dimensions as :attr:`X`

To me, that's the only API promise we're making here so we don't need to restructure the classes. I think this whole module is essentially internal, and any docs would be devdocs.

> but I’m tending towards my solution.

I'm not entirely against changing this module. I just think I'm pretty confident it works right now and would like to make a release. I'm not convinced of what changes would be the best, and would rather not have that decision hold a release up. It's also a little hard to see all of the consequences of this change in a PR with so many other changes.

Would you be okay with splitting this out of this PR and we can look at it after a release?

flying-sheep commented:
> Alright, I think I've done a full pass (though github keep changing the number of comment's it says I left, so maybe some things are repeated or left out). Most importantly: have you run the scanpy test suite on this branch?

Sure, every commit passes

> In general, I would like to see more information about why changes are being made inside of commit messages. There's a lot that changed here, and it wasn't always obvious to me why those changes happened. There are also many times when a change seems to have little or nothing to do with the rest of the commit. I don't think you have to go back through and split things out into more commits, since that's a pain, but in future it would make reviewing easier.

Hmm, apart from lumping the `AnnData.isview→is_view` rename and `isbacked` deprecation into The aligned_array commit, I don’t think so. I split everything up now, isn’t much work.

> For splitting code into new files, please be explicit about when code moves and when it get's changed. It's hard to tell if meaningful changes happened before and after code movement between files since github considers all lines as changing when they move.

Yeah, it’s a pain of some GitHub views, since it clearly tracks changes across files sometimes (e.g. in blame view) You simply mean to make two commits out of it, right?

flying-sheep commented on anndata/core/sparsedataset.py:
```diff
@@ -8,79 +8,117 @@
 
 
-class BackedSparseMatrixMixin:
+class BackedSparseMatrix(_cs_matrix):
```
Your concerns have always been addressed 11-15 lines above.

ivirshup replied:
I still think we should not explicitly inherit from a private class. Someone might start using a method from this class and we wouldn't know until it broke. Also, having been looking at the code for the class we'd be inheriting from, I want to distance ourself from it.

Are there other solutions to the problem being solved? Also, what exactly is the problem being solved?

* Can this be done in a stub file?
* Could this be an abstract base class with class attributes which help?

flying-sheep replied:
That’s the base class of `csr_matrix` and `csc_matrix`, we’re inheriting from that anyway farther down.

ivirshup replied:
Yeah, but there can be strange metaclasses in numpy code, and I don't think anything in the implementation of `BackedSparseMatrix` should even have the chance of relying on methods from `_cs_matrix`. If it inherits from that class, someone might try and use a method, which means that this would break with any scipy release that changes code (or renames a private class, or moves where it's defined) which has no API promises. I think it's better to head that off, since it was pretty painful for me to deal with before.

I know it's kind of doing that already by patching private methods, but I'd like to limit the reliance as much as possible.

flying-sheep replied:
OK!

flying-sheep commented on anndata/core/sparsedataset.py:
```diff
@@ -8,79 +8,117 @@
         else:
             return super().copy()
 
+    def _set_many(self, i: Iterable[int], j: Iterable[int], x):
```
No, they have been like this because that’s the way they were defined and attached to the class in scipy. The lines below like `backed_csr_matrix._zero_many = _zero_many` mean that the only difference to defining them like this was a different `__qualname__` and one more indentation level.

flying-sheep commented on anndata/core/sparsedataset.py:
```diff
@@ -130,205 +168,148 @@ def _get_sliceXarray(self, row, col):
 
 
-def get_format_str(data):
+def slice_len(s: slice, l: int) -> int:
```
No idea how! Didn’t know this was possible.

I always thought that git calculates diffs on the fly and there’s no way to change this apart from smartly doing two commits in a way that every individual one shows a more meaningful diff.

flying-sheep replied:
Which I did.

flying-sheep commented on anndata/core/views.py:
```diff
@@ -7,12 +7,13 @@
 from pandas.api.types import is_bool_dtype
 from scipy import sparse
 
+from . import anndata
```
No, else we get circular import problems.

flying-sheep commented on anndata/core/raw.py:
```diff
@@ -0,0 +1,218 @@
+from typing import Union, Mapping, Sequence
```
Things are getting huge. Going forward, I’d like to see e.g. all file-interaction abstracted to a submodule, with no abstractions leaking into `anndata.py` (e.g. `self.file` should probably never appear in that module)

An antidote to forgetting is to document it and other classes.

flying-sheep replied:
Tests remind you and as said: Raw should be documented in some fashion.

flying-sheep commented on anndata/core/anndata.py:
```diff
@@ -489,38 +489,27 @@ def _init_as_actual(
+            from ..readwrite.h5ad import _read_raw, _read_legacy_raw_h5ad
+
+            # Read raw dict from file
+            raw, raw_override = {}, raw
```
Hmm, I did my best to do something useful here. If we should throw more errors instead and do less guessing / creative interaction, sure!

ivirshup replied:
What did old versions do in this case?

flying-sheep replied:
Accept `None`, `Raw` instances, and dicts.

In case the anndata object was backed, there was some complex code for comparing `raw/X`’s shape with the shape of the `X` in the dict. Then the whole thing would be replaced …

My code now simply retrieves the backed Raw as dict and potentially overrides entries in that dict with user-specified dict members.

ivirshup replied:
> In case the anndata object was backed, there was some complex code for comparing raw/X’s shape with the shape of the X in the dict. Then the whole thing would be replaced …

Yeah, that was a bit of a nightmare.

> My code now simply retrieves the backed Raw as dict and potentially overrides entries in that dict with user-specified dict members.

There aren't a ton of tests around `.raw`. Could you make sure that cases this effects are covered, and if not, that they behave the same as they currently do?

It'd also be good to run some benchmarks (https://github.com/ivirshup/anndata-benchmarks) against this and see if it causes any regressions. There was a lot of unintended loading and copying during initialization before – especially for backed mode – which this suite helped catch.

flying-sheep replied:
Your benchmark suite should be integrated into the main tests! I don’t really have time to set everything up though at the moment. Any way this could possibly make performance worse?

ivirshup replied:
> Your benchmark suite should be integrated into the main tests!

Benchmarks run on travis won't have meaningful results, we've got to have somewhere to run them before they can be part of CI. Plus they need a little cleaning up.

> Any way this could possibly make performance worse?

There have been a lot of changes around instantiation, which is where we had lots of unnecessary copies being made and stored before, so yes.

A lot of the performance implications can also be non-obvious from reading the code.

flying-sheep commented on anndata/__init__.py:
```diff
@@ -10,9 +11,9 @@
-)  # backwards compat / shortcut for default format
+
+# backwards compat / shortcut for default format
+from .readwrite import read_h5ad as read
```
I just changed the comment here…

ivirshup replied:
This was more of a "while you're at it"

flying-sheep commented on anndata/core/sparsedataset.py:
```diff
@@ -130,205 +168,148 @@ def _get_sliceXarray(self, row, col):
+        mock_matrix = self.to_backed()
+        mock_matrix[row, col] = value
+
+    def _slice(
```
I just moved this…

flying-sheep replied:
Ah yeah, whoops, I removed duplication here.

ivirshup replied:
But could it be renamed to something like `_normalize_index` (which fits what I think it's doing)?

flying-sheep replied:
Done

flying-sheep commented on anndata/core/index.py:
```diff
@@ -0,0 +1,129 @@
+    elif isinstance(idx, (int, np.int_, np.int64, np.int32)):
+        return 1
+    else:
+        return len(idx)
```
Which one? please use the cool new “comment on line range” feature so I see the function name.

ivirshup replied:
In future, sure. But it's `get_n_items_idx`.

flying-sheep replied:
True it isn’t! Let’s get rid of it.

flying-sheep commented on anndata/readwrite/zarr.py:
```diff
@@ -252,23 +254,17 @@ def read_zarr(store: Union[str, Path, MutableMapping, zarr.Group]) -> AnnData:
-    if len(raw) > 0:
-        assert "raw" not in d
-        d["raw"] = raw
+    _read_legacy_raw_zarr(f, d.setdefault("raw", {}))
```
Won’t it be skipped later if it’s empty?

flying-sheep replied:
> If it is there, there would be a var

sure? Why?

ivirshup replied:
I was trying to think of what `skipped` means. I was thinking it could be when you're reading backed, (so there's an `X`, but it's not read).

Could you explain what you mean by "skipped"?

flying-sheep replied:
`_init_as_actual` contains

```py
if raw and isinstance(raw, cabc.Mapping):
    ...
elif raw:  # is a Raw from another AnnData
    ...
self._raw = raw if raw else None
```

so if `raw == {}`, `self._raw` will be set to `None`.

ivirshup replied:
Alright. I'm a little worried about what's inside those branches, but if it works it's fine. Could you take a quick look over the tests to make sure any case this could effect is checked?

flying-sheep commented on anndata/tests/test_base.py:
```diff
@@ -23,7 +23,7 @@
     AnnData(np.array([[1, 2], [3, 4]]))
     AnnData(np.array([[1, 2], [3, 4]]), {}, {})
-    AnnData(ma.array([[1, 2], [3, 4]]), uns={'mask': [0, 1, 1, 0]})
+    AnnData(ma.array([[1, 2], [3, 4]]), uns=dict(mask=[0, 1, 1, 0]))
```
Small driveby fix, sorry. I tried to be good mainly.

flying-sheep commented on anndata/tests/test_base.py:
```diff
@@ -42,7 +42,7 @@ def test_creation():
     # init with empty data matrix
     shape = (3, 5)
-    adata = AnnData(None, uns={'test': np.array((3, 3))}, shape=shape)
+    adata = AnnData(None, uns=dict(test=np.array((3, 3))), shape=shape)
```
I think that anything that’s really a kwarg in disguise shouls be done by `dict` to signify this.

flying-sheep replied:
Oh, right, this one is. Seems like this is more of a reflex now. Welp, no harm done.

ivirshup replied:
Alright, I was mainly curious. I felt like I saw this happening a bunch and wasn't sure if it was some style guide or something.

ivirshup commented:
So, if the commit hashes get invalidated, when I try to look at the code I left comments on I get this:

![image](https://user-images.githubusercontent.com/8238804/70017707-e279b380-15d7-11ea-93df-5ee1a19273a1.png)

Which makes reviewing take longer, since I now need to find the code and figure out how the history changed. Please don't rebase once a review has started.

ivirshup commented on anndata/core/aligned_mapping.py:
```diff
@@ -45,15 +43,32 @@ class AlignedMapping(cabc.MutableMapping, ABC):
+        else:  # Initialize view
+            assert subset_idx is not None, "A view needs a subset_idx"
+            return super().__new__(cls._View)
+
```
As mentioned in https://github.com/theislab/anndata/pull/260/files/47a688c0dbf59c488eaec40a30f1d462f6d75377#r352968031, I think this should be reverted.

flying-sheep replied:
answered there

ivirshup commented on anndata/core/sparsedataset.py:
```diff
@@ -8,79 +8,117 @@
+import collections.abc as cabc
 from itertools import accumulate, chain
-from typing import Union, NamedTuple, Tuple, Sequence, Type
+from typing import Union, NamedTuple, Tuple, Sequence, Iterable, Type
```
Why this? ABC classes seem like a more useful thing to use for typing.

flying-sheep replied:
I don’t get this:

![grafik](https://user-images.githubusercontent.com/291575/70048320-4dbc8900-15ca-11ea-87d7-478e10916156.png)

ivirshup replied:
The import changed:

```python
# Previously:
from collections.abc import Iterable
# Now:
from typing import Iterable
```

I didn't see why. A similar thing happened in `h5ad.py` with `Mapping`.

flying-sheep replied:
We’re using `from typing import ...` everywhere. `Iterable` exists in `typing`. `isinstance(x, collections.abc.Iterable)` works, `isinstance(x, typing.Iterable)` doesn’t necessarily.

For consistency we should keep `from typing import ...` and use `import collections.abc as cabc` to access `isinstance`-checkable ABCs.

ivirshup replied:
Hmm. My preference would probably be the other way, since the ABCs can be used in more cases. Use `import typing as typ` or something for brevity, but keep the most useful cases the shortest.

ivirshup commented on anndata/core/anndata.py:
```diff
@@ -515,7 +507,7 @@ def _init_as_actual(
 
         # layers
-        self._layers = Layers(self, layers)
+        self._layers = Layers(self, vals=(layers or {}))
```
If the case you're trying to deal with is `layers = None`, use an explicit check since other values can coerce to false.

e.g.

```python
self._layers = Layers(self, vals=(layers if layers is not None else {}))
```

flying-sheep replied:
That’s deliberate. We don’t care if one empty dict or another is used. As long as we don’t mutate them, the effect is the same.

flying-sheep replied:
answered above

ivirshup replied:
My thought was that it should throw an error if it's a type that shouldn't be here, like if it was `False`.

But I had forgotten there's already a solution to this. This should be `self._layers = Layers(self, vals=convert_to_dict(layers))` like the others (above):

```python
self._obsm = AxisArrays(self, 0, vals=convert_to_dict(obsm))
self._varm = AxisArrays(self, 1, vals=convert_to_dict(varm))

self._obsp = PairwiseArrays(self, 0, vals=convert_to_dict(obsp))
self._varp = PairwiseArrays(self, 1, vals=convert_to_dict(varp))
```

flying-sheep replied:
OK!

ivirshup commented on anndata/core/aligned_mapping.py:
```diff
@@ -329,24 +363,25 @@ def dim(self) -> str:
         self._parent = parent
-        if axis not in (0, 1):
-            raise ValueError()
+        assert axis in (0, 1)
```
Why this change?

flying-sheep replied:
Again: `asserts` are for us misusing internal interfaces or violating assumptions. “Not your fault user, but ours. Sorry and please file a bug”

At the moment this is an internal interface. We can change it back (with an actual helpful message) if it becomes a public one.

flying-sheep replied:
answered above

ivirshup commented on anndata/tests/test_layers.py:
```diff
@@ -62,6 +62,20 @@ def test_set_dataframe(homogenous, df, dtype):
+    assert np.allclose(layers.parent.X, adata.X)
+    assert layers.keys() == adata.layers.keys()
+    for new, old in zip(layers.values(), adata.layers.values()):
+        assert np.allclose(new, old)
```
Hmm. Any reason this isn't tested for the other types? Or for views?

And that `np.allclose` should be exact equality.

flying-sheep replied:
It’s exemplary. If one works, the others do too. It’s to see if `__getnewargs_ex__` works.

ivirshup replied:
Yeah, but I was pretty surprised by how many things failed when I started parameterizing tests by sparse matrix type. Previously anndata (pre-6.21 I think) did not do well with CSC matrices.

Just for safety, I think it's worth chucking a fixture on this.

flying-sheep commented:
> Which makes reviewing take longer, since I now need to find the code and figure out how the history changed. Please don't rebase once a review has started.

> Thanks, but also please don't rewrite history when other people are looking at it.

I think you missed that I’m agreeing with you! See my first 2 paragraphs in https://github.com/theislab/anndata/pull/260#issuecomment-560708450

flying-sheep commented on anndata/tests/test_hdf5_backing.py:
```diff
@@ -90,23 +90,23 @@ def test_read_write_X(tmp_path, mtx_format, backed_mode, force_dense):
 # this is very similar to the views test
 def test_backing(adata, tmp_path, backing_h5ad):
-    assert not adata.isbacked
+    assert not adata.filename
```
In that case “filename” is wrong too, and we should go to `backing: Union[Path, UnknownFutureObjectStore, None]`.

ivirshup replied:
In principle I agree with that, though I'd take this further and say this should be expressed at a type level. Unfortunately I don't think we can actually do that (since you'd need to change the type of an object).

Just for ease of use, I like functions like `is_*` for convenience in the API. `if adata.is_backed` is a little nicer than `if adata.backed is not None` (since `if val` can throw an error).

> I’m allergic to mutually exclusive properties (like if x is None y needs to be set): That stuff is untestable. Rather there should be a single enum property that can take all possible states.

I'd probably go for `adata.store: Union[AnnDataFileManager, ...]`, then the filename could be accessed from there. That said, I'd like to have some time playing around with what we can do with backed mode before making strong choices about an API.

flying-sheep replied:
`if adata.backing` or `if adata.store` is even shorter than `if adata.is_backed`.

`is not None` is only necessary if other falsy values should be treated differently than `None`.

ivirshup replied:
> if adata.backing or if adata.store is even shorter than if adata.is_backed

Yes, but they could be ambiguous. `is_backed` is like the `pd.api.type.is_*`, should be completely unambiguous.

----------------

> is not None is only necessary if other falsy values should be treated differently than None.

Which could happen.

To quote from the [google python style guide](http://google.github.io/styleguide/pyguide.html#214-truefalse-evaluations):

> Always use if foo is None: (or is not None) to check for a None value-e.g., when testing whether a variable or argument that defaults to None was set to some other value. The other value might be a value that’s false in a boolean context!

--------------------

> Are there APIs in AnnDataFileManager that should be exposed? Else I’d go with store/backing: Union[Path, Any, None] (the Any for future extensibility)

Maybe? I think backed mode implementation needs to be considered, and I'd wait on API changes until that happened.

flying-sheep commented on anndata/core/anndata.py:
```diff
@@ -1439,10 +1428,13 @@ def copy(self, filename: Optional[PathLike] = None) -> 'AnnData':
                 dtype=dtype,
             )
+            if self._raw is not None:
+                adata._raw = self._raw.copy()
```
Out of scope. There hasn’t been any before. Let’s do this after this PR

ivirshup replied:
?

It was passed through the AnnData constructor before, I think it should still do that. There is validation that happens there.

flying-sheep replied:
Ah, whoops, that’s the `copy` one not the `__init__` one.

That makes validation completely useless, since `adata` is a copy of `self`, so everything should be the same.

ivirshup replied:
Should be the same, but we also should check. Better this throws an error than doesn't.

flying-sheep commented:
OK ready apart from a few open conversations:

- AlignedMapping redesign https://github.com/theislab/anndata/pull/260#discussion_r352265463

  disposition: I think my design should be kept

- `_read_legacy_raw_zarr` and `setdefault` https://github.com/theislab/anndata/pull/260#discussion_r352269536

  disposition: `_read_legacy_raw_zarr` can go, but I don’t understand why `setdefault` should result in different behavior than leaving the key out

- `isbacked` vs `filename` vs `backing` https://github.com/theislab/anndata/pull/260#discussion_r352269771

  disposition: rename `filename` to `backing`, document that it might store something other than a filename in the future

- How should Raw initialization behave? https://github.com/theislab/anndata/pull/260#discussion_r352270961

- ABC classes (featuring a commit that is attributed to the wrong line) https://github.com/theislab/anndata/pull/260#discussion_r352978137

ivirshup commented:
> AlignedMapping redesign

I don't currently get the value of the changes you've made. I'd rather revert this and maybe add it later. I think it might be easier for me to understand these changes in a smaller more focused PR. I don't think these changes are affecting behavior, so I don't think they need to happen before a release.

> I think you missed that I’m agreeing with you!

I thought you were, but then there was a force push after that comment and broken links. I wasn't sure if I had understood then. Good to know we're on the same page now.

> isbacked vs filename vs backing

My preference would be to normalize the names (i.e. `isbacked` -> `is_backed`) but leave other API changes until we can a deeper look at backed mode. It's possible that we can make backed mode work better, but not with the current API and I'd like to figure that out before changing the API (just so it stays in one stable state for longer).

> _read_legacy_raw_zarr and setdefault

I don't actually mind `_read_legacy_raw_zarr` to much now. Have you added tests here against legacy zarr files?

For `setdefault`, I'm just not sure why the change is happening, and generally think if there isn't a practical effect the code shouldn't change.

ivirshup commented:
@flying-sheep, I've tried to run the benchmarks for this, but ran into a strange problem.

If I run `python setup.py build` on this branch, none of the files under `core/` get included in the build. Any idea what's up with that?

flying-sheep commented:
Yes. Also venting incoming:

Why can’t tools just use the correct way to detect Python modules? Why are they still relying on `__init__.py` which has been optional for 10 years?

flying-sheep commented:
Abandoning towards individual PRs, starting with #268
